File: MvvmSample/app/src/main/java/com/abben/mvvmsample/network/ApiResponse.java
package com.abben.mvvmsample.network;

import android.support.annotation.Nullable;
import android.util.Log;

import java.io.IOException;

import retrofit2.Response;

/**
 * 初始封装Retrofit返回的数据
 * Created by abben on 2017/10/25.
 */
public class ApiResponse<T> {

    public final int code;
    @Nullable
    public final T body;
    @Nullable
    public final String errorMessage;

    public ApiResponse(Throwable error) {
        code = 500;
        body = null;
        errorMessage = error.getMessage();
    }

    public ApiResponse(Response<T> response) {
        code = response.code();
        if(response.isSuccessful()) {
            body = response.body();
            errorMessage = null;
        } else {
            String message = null;
            if (response.errorBody() != null) {
                try {
                    message = response.errorBody().string();
                } catch (IOException ignored) {
                    Log.e("error", "ApiResponse: error while parsing response");
                }
            }
            if (message == null || message.trim().length() == 0) {
                message = response.message();
            }
            errorMessage = message;
            body = null;
        }
    }

    public boolean isSuccessful() {
        return code >= 200 && code < 300;
    }
}

File: MvvmSampleJetpack/app/src/main/java/com/abben/mvvmsamplejetpack/network/ApiResponse.kt
package com.abben.mvvmsamplejetpack.network

import retrofit2.Response

/**
 *  @author abben
 *  on 2019/1/18
 */
sealed class ApiResponse<T> {

    companion object {
        fun <T> create(error: Throwable): ApiErrorResponse<T> {
            return ApiErrorResponse(error.message ?: "unknown error")
        }

        fun <T> create(response: Response<T>): ApiResponse<T> {
            return if (response.isSuccessful) {
                if (response.body() == null || response.code() == 204) {
                    ApiEmptyResponse()
                } else {
                    ApiSuccessResponse(response.code(), response.body()!!)
                }
            } else {
                val errorMessage = response.errorBody()?.string()
                ApiErrorResponse(errorMessage ?: "unknown error")
            }
        }
    }
}

/**
 * 当http状态码为204时候为ApiSuccessResponse的body不为空而出现的类
 */
class ApiEmptyResponse<T> : ApiResponse<T>()

data class ApiSuccessResponse<T>(
    val code: Int,
    val body: T
) : ApiResponse<T>()

data class ApiErrorResponse<T>(val errorMessage: String) : ApiResponse<T>()
